### PR TITLE
Add APIs for ETCD Encryption

### PR DESCRIPTION
**Pull request by abhinavmpandey08**
*Description of changes:*
This PR adds APIs to support ETCD Encryption

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by pokearu — APPROVED** [review]
/lgtm

File: pkg/api/v1alpha1/etcdencryption_types.go [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+	Providers []EtcdEncryptionProvider `json:"providers"`
+	// Resources defines a list of objects and custom resources definitions that should be encrypted.
+	Resources []string `json:"resources"`
+}
+
+// EtcdEncryptionProvider defines the configuration for ETCD encryption providers.
+// Currently only KMS provider is supported.
+type EtcdEncryptionProvider struct {
+	// KMS defines the configuration for KMS Encryption provider.
+	KMS KMS `json:"kms"`
```
Comment (g-gaston) [inline pkg/api/v1alpha1/etcdencryption_types.go:14]: this probably need to be a pointer if we ever going to support different providers
So one can be filled and the other ones can be nill

File: pkg/api/v1alpha1/etcdencryption_test.go
Comment (g-gaston) [inline pkg/api/v1alpha1/etcdencryption_test.go:1]: remove file if we don't need tests here?

File: pkg/api/v1alpha1/etcdencryption_types.go [hunk 2]
```diff
@@ -0,0 +1,21 @@
+package v1alpha1
+
+// EtcdEncryption defines the configuration for ETCD encryption.
+type EtcdEncryption struct {
+	Providers []EtcdEncryptionProvider `json:"providers"`
+	// Resources defines a list of objects and custom resources definitions that should be encrypted.
+	Resources []string `json:"resources"`
```
Comment (g-gaston) [inline pkg/api/v1alpha1/etcdencryption_types.go:7]: question about this bc I don't know how it works
would all resources we encrypted with all providers?

File: pkg/api/v1alpha1/cluster_webhook.go [hunk 3]
```diff
@@ -103,6 +107,10 @@ func (r *Cluster) ValidateUpdate(old runtime.Object) error {
 
 	allErrs = append(allErrs, ValidateWorkerKubernetesVersionSkew(r, oldCluster)...)
 
+	if err := validateEtcdEncryptionConfig(r.Spec.EtcdEncryption); err != nil {
```
Comment (g-gaston) [inline pkg/api/v1alpha1/cluster_webhook.go:110]: do we need these validations in the CLI as well?

File: pkg/api/v1alpha1/etcdencryption_types.go (hunk 1 above)
Comment (abhinavmpandey08) [reply pkg/api/v1alpha1/etcdencryption_types.go:14]: We don't have plans to support another encryption provider besides KMS. But I can it a pointer, shouldn't make too much difference.

File: pkg/api/v1alpha1/etcdencryption_types.go (hunk 2 above)
Comment (abhinavmpandey08) [reply pkg/api/v1alpha1/etcdencryption_types.go:7]: You can have something like this where you configure one provider for some resources and another provider for other resources. 

```yaml
spec:
  etcdEncryption:
  - providers:
    - kms:
        socketListenAddress: unix:///var/run/kmsplugin/socket1-new.sock
    - kms:
        socketListenAddress: unix:///var/run/kmsplugin/socket1-old.sock
    resources:
    - secrets
    - custom-resource-1.anywhere.eks.amazonsaws.com
  - providers:
    - kms:
        socketListenAddress: unix:///var/run/kmsplugin/socket2-new.sock
    - kms:
        socketListenAddress: unix:///var/run/kmsplugin/socket2-old.sock
    resources:
    - configmaps
    - custom-resource-2.anywhere.eks.amazonsaws.com
```

In this example, `providers[0]` will be used to encrypt `secrets` and `custom-resource-1.anywhere.eks.amazonsaws.com` while `providers[1]` will be used to `configmaps` and `custom-resource-2.anywhere.eks.amazonsaws.com`

File: pkg/api/v1alpha1/cluster_webhook.go (hunk 3 above)
Comment (abhinavmpandey08) [reply pkg/api/v1alpha1/cluster_webhook.go:110]: yeah, I will add in this and more validations the CLI as well. I was thinking I would do it in a follow-up PR when I make other CLI changes. Does that sound good?

**Review by pokearu — APPROVED** [review]
/lgtm

**Pull request merged**